ci/docker_cache.py

```diff
@@ -0,0 +1,267 @@
+    """
+    try:
+        if S3_METADATA_IMAGE_ID_KEY in s3_object.metadata:
+            cached_image_id = s3_object.metadata[S3_METADATA_IMAGE_ID_KEY]
+            return cached_image_id
+        else:
+            logging.debug('No cached image available for {}'.format(s3_object.key))
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == "404":
+            logging.debug('{} does not exist in S3 yet'.format(s3_object.key))
+        else:
+            raise
+
+    return None
+
+
+def load_docker_cache(bucket_name, docker_tag):
+    # Allow anonymous access
+    s3_resource = boto3.resource('s3')
+    s3_resource.meta.client.meta.events.register('choose-signer.s3.*', disable_signing)
+    s3_object = s3_resource.Object(bucket_name, docker_tag)
+
+    # Check if cache is still valid and exists
+    remote_image_id = _get_remote_image_id(s3_object)
+    if remote_image_id:
+        if _docker_layer_exists(remote_image_id):
+            logging.info('Local docker cache already present for {}'.format(docker_tag))
+            return
+        else:
+            logging.info('Local docker cache not present for {}'.format(docker_tag))
+
+        # Download using public S3 endpoint (without requiring credentials)
+        with tempfile.TemporaryDirectory() as temp_dir:
+            tar_file_path = os.path.join(temp_dir, 'layers.tar')
+            s3_object.download_file(
+                Filename=tar_file_path,
+                Callback=ProgressPercentage(object_name=docker_tag, size=s3_object.content_length))
+
+            # Load layers
+            cmd = ['docker', 'load', '-i', tar_file_path]
+            try:
+                check_call(cmd)
+                logging.info('Docker cache for {} loaded successfully'.format(docker_tag))
+            except CalledProcessError as e:
+                logging.error('Error during load of docker cache for {} at {}'.format(docker_tag, tar_file_path))
+                logging.exception(e)
+                return
+    else:
+        logging.info('No cached remote image of {} present'.format(docker_tag))
+
+
+def _docker_layer_exists(layer_id):
+    """
+    Check if the docker cache contains the layer with the passed id
+    :param layer_id: layer id
+    :return: True if exists, False otherwise
+    """
+    cmd = ['docker', 'images', '-q']
+    image_ids_b = subprocess.check_output(cmd)
+    image_ids_str = image_ids_b.decode('utf-8').strip()
+    return layer_id in [id.strip() for id in image_ids_str.split('\n')]
+
+def _get_aws_session() -> boto3.Session:  # pragma: no cover
+    """
+    Get the boto3 AWS session
+    :return: Session object
+    """
+    global cached_aws_session
+    if cached_aws_session:
+        return cached_aws_session
+
+    session = boto3.Session()  # Uses IAM user credentials
+    cached_aws_session = session
+    return session
+
+def _format_docker_cache_filepath(output_dir, docker_tag):
```

Comment:
missing return type here for consistency

Reply:
Done